File: binary_tree/isSymmetric.cpp
// leetcode 101. isSymmetric cpp

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
 
class Solution {
public:
    bool isSymmetric(TreeNode* root) 
    {
        return this->dfs(root->left, root->right);
    }

    bool dfs(TreeNode* left, TreeNode* right)
    {
        if(!left && !right)
        {
            return true;
        }
        if(!left || !right)
        {
            return false;
        }
        if(left->val == right->val)
        {
            return this->dfs(left->left, right->right) && this->dfs(left->right, right->left);
        }
        return false;
    }
};

File: leetcode_hot_100/isSymmetric.py
# hot100-lc 101. 对称二叉树

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution(object):
    def isSymmetric(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        return self.dfs(root.left, root.right)

    def dfs(self, left, right):
        if not left and not right:
            return True
        if not left or not right:
            return False
        if left.val == right.val:
            return self.dfs(left.left, right.right) and self.dfs(left.right, right.left)
